File: src/main/java/at/yawk/fiction/impl/fanfiction/FfnFictionProvider.java
package at.yawk.fiction.impl.fanfiction;

import at.yawk.fiction.*;
import at.yawk.fiction.impl.PageParserProvider;
import java.net.URI;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;

/**
 * @author yawkat
 */
@RequiredArgsConstructor
public class FfnFictionProvider implements FictionProvider {
    private final PageParserProvider pageParserProvider;
    private final HttpClient httpClient;

    @Override
    public void fetchStory(Story story) throws Exception {
        Chapter chapter;
        if (story.getChapters() != null && !story.getChapters().isEmpty()) {
            chapter = story.getChapters().get(0);
        } else {
            chapter = new FfnChapter();
            ((FfnChapter) chapter).setIndex(0);
        }
        fetchChapter(story, chapter);
    }

    @Override
    public void fetchChapter(Story story, Chapter chapter) throws Exception {
        pageParserProvider.getParser(ChapterPageParser.class).request(httpClient)
                .get("https://www.fanfiction.net/s/" + ((FfnStory) story).getId() + '/' +
                     (((FfnChapter) chapter).getIndex() + 1))
                .target(new ChapterPageParser.StoryChapterPair(story, chapter))
                .send();
    }

    @Override
    public FfnSearchQuery createQuery() {
        return new FfnSearchQuery();
    }

    public List<FfnSubCategory> fetchSubCategories(FfnCategory category) throws Exception {
        List<FfnSubCategory> subCategories = pageParserProvider.getParser(SubCategoryPageParser.class)
                .parse(httpClient, new HttpGet("https://www.fanfiction.net/" + category.getId()));
        for (FfnSubCategory subCategory : subCategories) {
            subCategory.setCategory(category);
        }
        return subCategories;
    }

    @Override
    public Pageable<FfnStory> search(SearchQuery query) {
        // fail early
        FfnSearchQuery ffnSearchQuery = (FfnSearchQuery) query;
        return page -> {
            URI uri = ffnSearchQuery.build(page);
            Pageable.Page<FfnStory> o = pageParserProvider.getParser(SearchPageParser.class)
                    .request(httpClient).get(uri).send();
            o.setLast(o.getPageCount() <= page - 1);
            return o;
        };
    }
}

File: src/main/kotlin/at/yawk/fiction/impl/fanfiction/FfnFictionProvider.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package at.yawk.fiction.impl.fanfiction

import at.yawk.fiction.*
import org.apache.http.client.HttpClient
import org.apache.http.client.methods.HttpGet

/**
 * @author yawkat
 */
class FfnFictionProvider(private val httpClient: HttpClient) : FictionProvider {
    override val providedTypes = listOf(
            FfnAuthorProviderData::class.java,
            FfnChapterProviderData::class.java,
            FfnStoryProviderData::class.java,
            FfnSearchQuery::class.java
    )

    override fun fetchStory(story: Story): Story {
        val chapter: Chapter
        if (story.chapters != null && !story.chapters.isEmpty()) {
            chapter = story.chapters[0]
        } else {
            chapter = Chapter(providerData = FfnChapterProviderData(index = 0))
        }
        return fetchChapter(story, chapter).first
    }

    override fun fetchChapter(story: Story, chapter: Chapter): Pair<Story, Chapter> {
        val id = (story.providerData as FfnStoryProviderData).id
        val i = (chapter.providerData as FfnChapterProviderData).index!! + 1
        return ChapterPageParser.request(httpClient)
                .get("https://www.fanfiction.net/s/$id/$i")
                .send()
    }

    fun fetchSubCategories(category: FfnCategory): List<FfnSubCategory> {
        val subCategories = SubCategoryPageParser.fetch(httpClient, HttpGet("https://www.fanfiction.net/" + category.id))
        return subCategories.map { it.copy(category = category) }
    }

    override fun search(query: SearchQuery): Pageable<Story> {
        // fail early
        val ffnSearchQuery = query as FfnSearchQuery
        return object : Pageable<Story> {
            override fun getPage(page: Int): Pageable.Page<Story> {
                val uri = ffnSearchQuery.build(page)
                val o = SearchPageParser.request(httpClient).get(uri).send()
                return o.copy(last = o.pageCount!! <= page - 1)
            }
        }
    }
}
